algebra/matrix.c

```diff
@@ -280,35 +280,38 @@ int matrix_add(matrix_t *A, matrix_t *B, matrix_t *C)
 	unsigned int row, col; /* represent position in output C matrix */
 
 	if (C != NULL) {
-		/* force switch C matrix to untransposed */
-		memset(C->data, 0, sizeof(float) * C->rows * C->cols);
-		/* force switch C matrix to untransposed */
-		if (C->transposed) {
-			row = C->rows;
-			C->rows = C->cols;
-			C->cols = row;
-			C->transposed = A->transposed;
+		if ((C->rows != A->rows || C->cols != A->cols) && (C->rows != A->cols && C->cols != A->rows)) {
```

Comment:
```suggestion
		if (!(C->rows == A->rows && C->cols == A->cols) || !(C->rows == A->cols && C->cols == A->rows)) {
```
Just a suggestion. You can leave it as it is, this feels more natural as we check if they: (don't have common sizes) or (don't have common sizes if some is transposed)

Reply:
Ok, but in your suggestion the middle operator should be `&&`